server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/SetMailboxesCreationProcessor.java

```diff
@@ -19,20 +19,130 @@
+            collectionHierarchySorter.sortFromRootToLeaf(request.getCreate().entrySet())
+                .forEach(entry -> 
+                    createMailbox(entry.getKey(), entry.getValue(), mailboxSession, creationIdsToCreatedMailboxId, builder));
+        } catch (CycleDetectedException e) {
+            request.getCreate().entrySet()
+                .forEach(entry ->
+                    builder.notCreated(entry.getKey(),
+                            SetError.builder()
+                            .type("invalidArguments")
+                            .description("The created mailboxes introduce a cycle.")
+                            .build()));
+        }
         return builder.build();
     }
 
+    private void createMailbox(MailboxCreationId mailboxCreationId, MailboxRequest mailboxRequest, MailboxSession mailboxSession,
+            Map<MailboxCreationId, String> creationIdsToCreatedMailboxId, SetMailboxesResponse.Builder builder) {
+        try {
+            checkName(mailboxRequest, mailboxSession);
+            MailboxPath mailboxPath = getMailboxPath(mailboxRequest, creationIdsToCreatedMailboxId, mailboxSession);
+            mailboxManager.createMailbox(mailboxPath, mailboxSession);
+            Optional<Mailbox> mailbox = mailboxUtils.mailboxFromMailboxPath(mailboxPath, mailboxSession);
+            if (mailbox.isPresent()) {
+                builder.creation(mailboxCreationId, mailbox.get());
+                creationIdsToCreatedMailboxId.put(mailboxCreationId, mailbox.get().getId());
+            } else {
+                builder.notCreated(mailboxCreationId, SetError.builder()
+                        .type("anErrorOccurred")
+                        .description("An error occurred when creating the mailbox")
+                        .build());
+            }
+        } catch (MailboxNameException | MailboxParentNotFoundException e) {
+            builder.notCreated(mailboxCreationId, SetError.builder()
+                    .type("invalidArguments")
+                    .description(e.getMessage())
+                    .build());
+        } catch (MailboxExistsException e) {
+            String message = String.format("The mailbox '%s' already exists.", mailboxCreationId.getCreationId());
+            builder.notCreated(mailboxCreationId, SetError.builder()
+                    .type("invalidArguments")
+                    .description(message)
+                    .build());
+        } catch (MailboxException e) {
+            String message = String.format("An error occurred when creating the mailbox '%s'", mailboxCreationId.getCreationId());
+            LOGGER.error(message, e);
+            builder.notCreated(mailboxCreationId, SetError.builder()
+                    .type("anErrorOccurred")
+                    .description(message)
+                    .build());
+        }
+    }
+
+    private void checkName(MailboxRequest mailboxRequest, MailboxSession mailboxSession) {
```

Comment:
"ensure" of "assert" are usually more explicit when implementing some sort of preconditions, I'd suggest.